src/add_context.rs

```diff
@@ -2,93 +2,85 @@
 //! to a wrapped service.
 
 use super::{Push, XSpanIdString};
+use context::ContextualPayload;
+use futures::Future;
 use hyper;
-use hyper::{Error, Request, Response};
-use std::io;
+use hyper::Request;
 use std::marker::PhantomData;
 
+use ErrorBound;
+
 /// Middleware wrapper service, that should be used as the outermost layer in a
 /// stack of hyper services. Adds a context to a plain `hyper::Request` that can be
 /// used by subsequent layers in the stack.
 #[derive(Debug)]
-pub struct AddContextNewService<T, C>
-where
-    C: Default + Push<XSpanIdString>,
-    T: hyper::server::NewService<
-        Request = (hyper::Request, C::Result),
-        Response = hyper::Response,
-        Error = hyper::Error,
-    >,
-{
```

Comment:
The type constraints force a lifetime to form part of the type, despite that lifetime not being constrained.

This is really tedious, because if an AddContextMakeService is embedded in a struct, then the lifetime percolates outwards indefinitely (i.e. all structs containing the type have no option but to also include a lifetime parameter).

Furthermore, while we don't have a use case at the moment, it'd be useful to be able to implement MakeService for both Service Contexts passed by reference, and those by value, and in that case, we can just add a second impl, with different constraints, but the same underlying type.

Reply:
Is that true for C as well as T?

Reply:
Otherwise fine.